src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/DirectoryBasedStoreProvider.cs

```diff
@@ -0,0 +1,454 @@
+                        dataExistsAlready = true;
+                    }
+
+                    if (dataExistsAlready)
+                    {
+                        // The file was added but our collection hasn't resynced yet.
+                        // Set _certificates to null to force a resync.
+                        _certificates = null;
+                        return;
+                    }
+                }
+
+                string destinationFilename;
+                FileMode mode = FileMode.CreateNew;
+
+                if (existingFilename != null)
+                {
+                    destinationFilename = existingFilename;
+                    mode = FileMode.Create;
+                }
+                else if (findOpenSlot)
+                {
+                    destinationFilename = FindOpenSlot(thumbprint);
+                }
+                else
+                {
+                    destinationFilename = Path.Combine(_storePath, thumbprint + PfxExtension);
+                }
+
+                using (FileStream stream = new FileStream(destinationFilename, mode))
+                {
+                    byte[] pkcs12 = copy.Export(X509ContentType.Pkcs12);
+                    stream.Write(pkcs12, 0, pkcs12.Length);
+                }
+
+#if DEBUG
+                // Verify that we're creating files with u+rw and o-rw, g-rw.
+                const Interop.libc.Permissions requiredPermissions =
+                    Interop.libc.Permissions.S_IRUSR |
+                    Interop.libc.Permissions.S_IWUSR;
+
+                const Interop.libc.Permissions forbiddenPermissions =
+                    Interop.libc.Permissions.S_IROTH |
+                    Interop.libc.Permissions.S_IWOTH |
+                    Interop.libc.Permissions.S_IRGRP |
+                    Interop.libc.Permissions.S_IWGRP;
+
+                Interop.Sys.FileStatus stat;
+
+                Debug.Assert(Interop.Sys.Stat(destinationFilename, out stat) == 0);
+                Debug.Assert((stat.Mode & (int)requiredPermissions) != 0, "Created PFX has insufficient permissions to function");
+                Debug.Assert((stat.Mode & (int)forbiddenPermissions) == 0, "Created PFX has too broad of permissions");
+#endif
+            }
+
+            // Null out _certificates so the next call to get_Certificates causes a re-scan.
+            _certificates = null;
+        }
+
+        public void Remove(ICertificatePal certPal)
+        {
+            OpenSslX509CertificateReader cert = (OpenSslX509CertificateReader)certPal;
+
+            using (X509Certificate2 copy = new X509Certificate2(cert.DuplicateHandles()))
+            {
+                bool hadCandidates;
+                string currentFilename = FindExistingFilename(copy, _storePath, out hadCandidates);
+
+                if (currentFilename != null)
+                {
+                    if (_readOnly)
```

Comment:
So, originally I had made the trivial tests of Add and Remove while ReadOnly.  Add passed on Windows (threw the exception), Remove didn't.  So I made the test with lots of comments about how the test can't really work happily, yadda yadda.  Didn't occur at all to me that "add when it already exists" would be a similar no-op.  But it is.  So, grr.  Making `Add` now only throw when it thinks it needs to do work.